systemd/cloud-init-single.service

```diff
@@ -0,0 +1,27 @@
+[Unit]
```

Comment:
> If these pieces are templatised there, don't they also need to be here?

Upstream cloud-init gets tested with a single systemd ordering. We've accepted these downstream contributions, but we don't actually test that they work. As it is today, we could very easily break a downstream with a change in upstream cloud-init, but because these alternate orderings don't get tested we would have no way of knowing. Attempt to support multiple orderings in upstream isn't honest given the current testing situation. I think that it would instead be better to try to implement unified units which are both correct and opinionated for all various distros.

My claim is that these templated changes either 1) don't need to be templated 2) are redundant/unnecessary or 3) are wrong.

I'm happy to be disproved, but since this new unit is a breaking change already anyways, it's an opportunity to implement what we believe is correct and require that if we really need to template these files that a good reason is provided besides "that's the way it was before".

To address the ones you mentioned:

> Before=firewalld.target

No need to template. Systemd won't block on non-existent services so we can add this right in.

> Conflicts=shutdown.target

No need to template - this should be the default everywhere. This is a default `Conflicts=` and `Before=` dependency for units except those that have `DefaultDependencies=no`.

> Before=sysinit.target

This is actually redundant because `sysinit.target` is ordered after `network-pre.target` - we shouldn't need it.

> ExecStartPre=/bin/mkdir -p /run/cloud-init

The responsibility for the existence of this directory belongs to the package manager, not the init system. [See here](https://stackoverflow.com/questions/3308785/installing-an-empty-directory-with-rpm) for a suggestion on how to do this in a spec file.

> ExecStartPre=/sbin/restorecon /run/cloud-init

The packaging should also be able to do this either with restorecon or alternatively with an selinux policy module. I don't see a compelling reason to require modifying this directory when it boots.

> ExecStartPre=/usr/bin/touch /run/cloud-init/enabled

Um... what? Cloud-init shouldn't be "always enabled". ds-identify enables cloud-init if it is found to be running on a suitable cloud. Downstreams can do whatever they want, but in upstream we should be opinionated about cloud-init is supposed to work. We don't need to be in the business of endorsing and supporting dozens of different cloud-init implementations that all behave slightly differently. Cloud-init should run when it detects that it is on a supported cloud - not always. This is how it is documented to work, and this is what we should encourage in downstreams.

Reply:
Aside: Since systemd doesn't calculate ordering for units that don't exist, I suspect that we could probably drop _all_ of the jinja templates and have a unified ordering that is the same for all distros, but that's way out of scope for this PR - for now I just want to only include what we know is needed rather than copy and paste tech debt.

Reply:
> > Attempt to support multiple orderings in upstream isn't honest given the current testing situation.
> 
> This is very extreme to me and I do not agree at all. Yes, testing needs to be improved in multiple areas, but just because something isn't tested doesn't mean there isn't an honest attempt at support. We have loads of things that are untested or that we rely on others to test for us, but we still aim to support. I'm thinking of things like renderers/activators not included in Ubuntu, other init systems, and BSD, just to name a few.

Yeah, honest wasn't the right word. What I was attempting to say is that I think that supporting multiple systemd orderings is expensive and messy. I'm challenging the idea that it is an effective use of upstream resources.

I think that we can probably eliminate most, if not all, of the templates while still working on downstreams - though a couple of tweaks may be necessary elsewhere. Systemd services should be re-usable across distros. Systemd makes this possible with  abstract synchronization points such that various classes of services (firewalls, networking daemons, desktop environments, etc) should be pluggable - one can be swapped out for another, and as long as they have the correct `After=`/`Before=` ordering things should "just work" for the most part from an ordering perspective.

> I'm thinking of things like renderers/activators not included in Ubuntu, other init systems, and BSD, just to name a few.

These aren't as impactful because they are all pluggable components which are much more insulated than the ordering of the entire init system.

Supporting N renderers/activators, package managers, init systems, etc, is relatively cheap, because they are designed as swappable components so a change to the netplan renderer shouldn't break the networkmanager one, for example. Individual components such as renderers/activators, package managers, etc are much more insulated from changes to cloud-init because they fall within the ordering of the entire system - they do not modify it except by possibly slowing boot for users of that component.

Many core cloud-init assumptions are dependent on the ordering of the instance's initialization and the availability of specific resources (fs/network/etc).

Attempting to modify/maintain a cloud-init that is supposed to support N systemd orderings is costly, because it means that any change to ordering within cloud-init carries the risk of breaking one of the many systemd ordering permutations. There is no way to limit the splash damage of ordering changes in cloud-init - they are a global problem across the entire codebase, since a modified ordering shifts when the entire codebase starts running.

An example of an obvious ordering change that carries outsized risk across different distros is [this one](https://github.com/canonical/cloud-init/commit/4981ea9a5ba3b5b267f1c4862501634adfc7f55e), since it moves a configuration module to an earlier stage in boot, which is ordered differently for different distros. Ordering changes within a stage could also cause disruption, yet would be far more subtle.

While these templated unit files make it easy for downstreams to upstream their code without disruption, it is an even greater maintenance burden for upstream than downstream since we don't have the history of the origin of these changes in upstream bug trackers or repositories, nor do we have the expertise or infrastructure to maintain and test the code for the distributions that these templates exist for. Additionally, I think that most (if not all) of these templated changes either a) should be defaults for all b) are redundant c) are unnecessary, or d) would be more appropriate to implemented elsewhere.

> Given that `/run` is ephemeral...  To your specific points, the package manager can't ensure that anything in `/run` exists on a subsequent boot, correct? 

Oops, you're right. Since that's the case, packaging isn't the right place, but I still strongly believe that a cloud-init unit file isn't the right place for it either.

> For `/run/cloud-init/enabled`, I don't think the touch is saying that cloud-init should always be enabled, and as far as I know that file isn't involved in determining whether cloud-init should run. If this line is run, the generator will have already run and enabled the target (and thus the unit files). If the generator disabled cloud-init, this line shouldn't run.

Ahh, true. So what is it even trying to do then? I don't see how this could do anything useful.

> To your broader point, it's probably a good idea to take it to the distro and see if
> 
>     1. There's a compelling reason it's implemented this way
> 
>     2. It makes sense to keep it in distro packaging
> 
> 
> but I don't think that the reasons you've given justifies breaking currently working distros. 
> These changes are [only two years old](https://github.com/canonical/cloud-init/pull/1431).
> > I'm happy to be disproved, but since this new unit is a breaking change already anyways, it's an opportunity to implement what we believe is correct and require that if we really need to template these files that a good reason is provided besides "that's the way it was before".
> 
> I thought the whole aim of this PR was to _not_ intentionally break anything.

I don't think that that is possible. By necessity, this change has to break something for somebody. The proposal was that stable downstreams can maintain backwards compatibility by patching out just the systemd changes from this PR, and that unpatched (stable / rolling) downstreams would be a reasonable place to do the systemd unit cleanup.

> Regardless, if we're going to make unrelated, intentionally breaking changes, I really think they need to go in a separate PR. If we want the behavior to be as close as possible as to what it was before, it seems this isn't an honest attempt to do that on a RHEL system.

I'm fine with pursuing a systemd unit file cleanup as a separate effort from the single process optimization - it will ease reviews if nothing else.

Reply:
>> Before=sysinit.target
>
> This is actually redundant because sysinit.target is ordered after network-pre.target - we shouldn't need it.

On further examination, I couldn't find any ordering dependency between network-pre.target and sysinit.target, but both appear in the critical chain because they are both required, so it is probably best to keep both.